### PR TITLE
Fix BlockEvent.PlaceEvent ignoring EnumHand (#3221)

### LINKED ISSUE #3221: BlockPlace's itemInHand does not respect off_hand
If we place a block using our off hand with nothing in our main hand the itemInHand value will be null.

There's not proper way to know which hand was used to place the block.

The event is created in ForgeEventFactory.onPlayerBlockPlace and is fired from ForgeHooks.onPlaceItemIntoWorld. That last method receives an EnumHand value, that could be passed to the event and it could be used to determine which ItemStack was used.

Forge version 12.18.1.2011 (MC 1.10.2)

### PR DESCRIPTION
**Pull request by Choonster**
This fixes #3221.

### CODE REVIEW

**Comment by Actuarius** [comment]
@mezz added labels [Bug]

**Comment by jredfox** [comment]
@Choonster 
I noticed in this code that You used the main hand and blockstate 0 for multiblock place. Are you sure you couldn't have offhand as well or would that be more work? I also noticed in previous files that blockstate had an if statement for blockstate 1 did you miss this for your mutliblock place method?

 this(blockSnapshots, placedAgainst, player, EnumHand.MAIN_HAND);
-        }
  +
-        public MultiPlaceEvent(List<BlockSnapshot> blockSnapshots, IBlockState placedAgainst, EntityPlayer player, EnumHand hand) {
-            super(blockSnapshots.get(0), placedAgainst, player, hand);

**Comment by Choonster** [comment]
I only use the main hand for the deprecated overloads of the event constructors and factory methods to preserve the old behaviour for mods that use them. The new overloads used by Forge take an `EnumHand` argument.

The new code uses `blocksSnapshots.get(0)` in exactly the same way as the old code: to get the snapshot when there's only one snapshot, to get the `placedAgainst` state when there's multiple snapshots and to pass the first snapshot to the `PlaceEvent` constructor from the `MultiPlaceEvent` constructor.

Just like before, mods that subscribe to `MultiPlaceEvent` will have access to all the snapshots (if there was more than one) and mods that subscribe to `PlaceEvent` will have access to the first snapshot.

**Comment by jredfox** [comment]
When Will this be implemented? I would like to get a working version of my mod done. It seems Like they checked it says no conflicts and it's been like a week.

**Comment by jredfox** [comment]
Don't mean to be rude but, if all checks have passed no conflicts with it why hasn't it been implemented? I patentially waited a week and saw a forge update got excited then realized that this fix wasn't in there.  **This is critical with doing any advanced mod coding.  So could you put this in the next version please and make this a priority**?  @LexManos

**Comment by gr8pefish** [comment]
I too would like to see this implemented. @Choonster has provided a good, clean implementation of a necessary addendum to the event.

**Comment by jredfox** [comment]
Well then it's agreed then that this is good clean and bug free then and the code is provided then forge now needs to add this in the next build. Mods won't be the same without this fix.

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 1]
```diff
@@ -112,6 +112,10 @@
 public class ForgeEventFactory
 {
 
+    /**
+     * @deprecated Use {@link #onPlayerMultiBlockPlace(EntityPlayer, List, EnumFacing, EnumHand)} instead.
+     */
+    @Deprecated
     public static MultiPlaceEvent onPlayerMultiBlockPlace(EntityPlayer player, List<BlockSnapshot> blockSnapshots, EnumFacing direction)
     {
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java:122]: Dont duplicate code, bounce to the new one with the argument null, this makes it more obvious what to remove when deleting deprecated functions. And illustrates the parameter CAN be null.

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 2]
```diff
@@ -129,6 +146,14 @@ public static PlaceEvent onPlayerBlockPlace(EntityPlayer player, BlockSnapshot b
         return event;
     }
 
+    public static PlaceEvent onPlayerBlockPlace(EntityPlayer player, BlockSnapshot blockSnapshot, EnumFacing direction, EnumHand hand)
+    {
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java:145]: Same as above

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 3]
```diff
@@ -170,18 +171,25 @@ public void setExpToDrop(int exp)
...
         public PlaceEvent(BlockSnapshot blockSnapshot, IBlockState placedAgainst, EntityPlayer player)
         {
+            this(blockSnapshot, placedAgainst, player, EnumHand.MAIN_HAND);
+        }
+
+        public PlaceEvent(BlockSnapshot blockSnapshot, IBlockState placedAgainst, EntityPlayer player, EnumHand hand) {
             super(blockSnapshot.getWorld(), blockSnapshot.getPos(), blockSnapshot.getCurrentBlock());
             this.player = player;
-            this.itemInHand = player.getHeldItemMainhand();
+            this.itemInHand = player.getHeldItem(hand);
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: You may need null protection here.

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 4]
```diff
@@ -190,6 +198,7 @@ public PlaceEvent(BlockSnapshot blockSnapshot, IBlockState placedAgainst, Entity
         public BlockSnapshot getBlockSnapshot() { return blockSnapshot; }
         public IBlockState getPlacedBlock() { return placedBlock; }
         public IBlockState getPlacedAgainst() { return placedAgainst; }
+        public EnumHand getHand() { return hand; }
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java:205]: @Nullable

**Comment by Choonster** [comment]
The new commit should address your comments.

**Comment by gr8pefish** [comment]
If anything else is needed, I am willing to help out in whatever regard I can to make this PR be pulled :)

**Comment by gr8pefish** [comment]
Any word on if this will be accepted? I just looked over the code myself and noted no issues with it; seeing as it addresses an issue I would like to see it merged in the near future if possible. Or at least an explanation of what is causing a holdup (if anything).

That being said, if it isn't ever going to be merged, I would simply like to know that so I can start writing my own workarounds. Thanks!

**Pull request merged**